### ISSUE TITLE
other exchange

### ISSUE THREAD

**Issue by hosseinab** [ChristopherBenjaminHemmens/bittrex-arbitrage#2]
hello
can i implement this project for other exchange ?

**Comment by ChristopherBenjaminHemmens** [comment]
You could but you'd need to implement the API

**Comment by hosseinab** [comment]
@ChristopherBenjaminHemmens hi . thank you for answer .
Have you seen ccxt library?
https://github.com/ccxt/ccxt (in this library all exchanges api implemented )
Is it possible to integrate this project with ccxt?

**Comment by ChristopherBenjaminHemmens** [comment]
I'm not sure to what extent it's possible to exchange data with the crypto exchanges in terms of using websockets rather than using polling in ccxt. Ccxt is heavily geared up towards web requests. You could build something that works no doubt, and it would probably be fairly standardized and easy to implement. But it would face performance issues if I can't use websockets I;d have to send loads of webrequests very frequently to get the market prices.

I had to copy Bittrex's example socket client code and modify the code to implement it into this project.

**Comment by hosseinab** [comment]
@ChristopherBenjaminHemmens 
Does this mean that the implementation of this project with ccxt is not possible?

**Comment by ChristopherBenjaminHemmens** [comment]
Well not fully. It means it's partially possible.

The models that require web requests from exchanges could be integrated using ccxt, so you could abstract a layer and plug ccxt into Account, Balance, Currency, Order and Trade models in the [application/model](https://github.com/ChristopherBenjaminHemmens/bittrex-arbitrage/tree/master/application/model)

Then you'd still need a way of implementing a socket in each exchange to get fast real-time data for Market. That's where the big issue would lie.

You'd still need to create a library, or extend ccxt with web socket or alternative socket functionality to prevent a large amount of web requests if you were to poll the exchanges for market prices.

**Comment by hosseinab** [comment]
@ChristopherBenjaminHemmens So, if the number of web requests is high, will there be a problem?

**Comment by ChristopherBenjaminHemmens** [comment]
Yes. You will have a high amount of data going over your network as a starting point. That will factor to more memory usage, and lower capacity for traffic through your network.

You will most likely face throttling issues. For example if I send more than three web requests to Bittrex withing one second is fine, but trying a fourth will respond in a throttling error. Most exchanges will be implemented like this to prevent being more or less DDOS attacked for their tickers.

If my machine is waiting on a response from a web request containing a lot of data, it would take more computing power to process all the routes one on chunk, rather than each route separately as when the data is served to me. Not only that as I'm not receiving the ticker prices synchronously from the exchange but rather waiting x amount of time to ask for the next ticker, the age of my prices could already be too old as someone could have already calculated the conflict before I even got the prices.

Then were I to request prices and send requests for three trades simultaneously I will most likely face throttling issues when placing the trades.

**Comment by ChristopherBenjaminHemmens** [comment]
I'm going to start implementing Binance to start with many exchanges. It will be quite a bit of work but it should be achievable. I'm going to select the exchanges one by one because while it's easy to use a library that has generic functionality it will have its limits. I'm going to use node-binance-api as this module already has websocket support implemented so that will aleviate a huge headache.

Once I have a generic system where it's possible to implement many exchanges in a certain structure, it should be quite easy to start implementing other exchanges, but it will take a while before the model is completely reprogrammed to work on many exchanges as the initial setup was one exchange.

**Comment by hosseinab** [comment]
@ChristopherBenjaminHemmens thank you . its great
